File: app/src/main/java/cn/cqautotest/sunnybeach/ui/activity/BrowserActivity.java
package cn.cqautotest.sunnybeach.ui.activity;

import android.app.Activity;
import android.content.Context;
import android.content.Intent;
import android.graphics.Bitmap;
import android.graphics.drawable.BitmapDrawable;
import android.text.TextUtils;
import android.view.KeyEvent;
import android.view.View;
import android.webkit.CookieManager;
import android.webkit.WebView;
import android.widget.ProgressBar;

import androidx.annotation.NonNull;

import com.hjq.umeng.Platform;
import com.hjq.umeng.UmengShare;
import com.scwang.smart.refresh.layout.SmartRefreshLayout;
import com.scwang.smart.refresh.layout.api.RefreshLayout;
import com.scwang.smart.refresh.layout.listener.OnRefreshListener;
import com.umeng.socialize.media.UMImage;
import com.umeng.socialize.media.UMWeb;

import java.util.List;

import cn.cqautotest.sunnybeach.R;
import cn.cqautotest.sunnybeach.action.StatusAction;
import cn.cqautotest.sunnybeach.aop.CheckNet;
import cn.cqautotest.sunnybeach.aop.DebugLog;
import cn.cqautotest.sunnybeach.app.AppActivity;
import cn.cqautotest.sunnybeach.app.AppApplication;
import cn.cqautotest.sunnybeach.db.CookieRoomDatabase;
import cn.cqautotest.sunnybeach.db.SobCacheManager;
import cn.cqautotest.sunnybeach.db.dao.CookieDao;
import cn.cqautotest.sunnybeach.manager.CookieStore;
import cn.cqautotest.sunnybeach.manager.ThreadPoolManager;
import cn.cqautotest.sunnybeach.other.FitScreen;
import cn.cqautotest.sunnybeach.other.IntentKey;
import cn.cqautotest.sunnybeach.ui.dialog.ShareDialog;
import cn.cqautotest.sunnybeach.util.Constants;
import cn.cqautotest.sunnybeach.util.StringUtil;
import cn.cqautotest.sunnybeach.widget.BrowserView;
import cn.cqautotest.sunnybeach.widget.StatusLayout;
import okhttp3.Cookie;
import timber.log.Timber;

/**
 * author : Android 轮子哥 & A Lonely Cat
 * github : https://github.com/getActivity/AndroidProject
 * time   : 2018/10/18
 * desc   : 浏览器界面
 */
public final class BrowserActivity extends AppActivity implements StatusAction, OnRefreshListener {

    @CheckNet
    @DebugLog
    public static void start(Context context, String url) {
        if (TextUtils.isEmpty(url)) {
            return;
        }
        Intent intent = new Intent(context, BrowserActivity.class);
        intent.putExtra(IntentKey.URL, url);
        // 加载非反馈界面
        intent.putExtra(IntentKey.OTHER, false);
        if (!(context instanceof Activity)) {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        }
        context.startActivity(intent);
    }

    @CheckNet
    @DebugLog
    public static void start(Context context, String url, String openId, String nickName, String avatar) {
        if (TextUtils.isEmpty(url)) {
            return;
        }
        Intent intent = new Intent(context, BrowserActivity.class);
        intent.putExtra(IntentKey.URL, url);
        intent.putExtra(IntentKey.ID, openId);
        intent.putExtra(IntentKey.NAME, nickName);
        intent.putExtra(IntentKey.AVATAR_URL, avatar);
        // 加载反馈界面
        intent.putExtra(IntentKey.OTHER, true);
        if (!(context instanceof Activity)) {
            intent.addFlags(Intent.FLAG_ACTIVITY_NEW_TASK);
        }
        context.startActivity(intent);
    }

    private StatusLayout mStatusLayout;
    private ProgressBar mProgressBar;
    private SmartRefreshLayout mRefreshLayout;
    private BrowserView mBrowserView;

    @Override
    protected int getLayoutId() {
        return R.layout.browser_activity;
    }

    @Override
    protected void initView() {
        mStatusLayout = findViewById(R.id.hl_browser_hint);
        mProgressBar = findViewById(R.id.pb_browser_progress);
        mRefreshLayout = findViewById(R.id.sl_browser_refresh);
        mBrowserView = findViewById(R.id.wv_browser_view);

        // 设置 WebView 生命管控
        mBrowserView.setLifecycleOwner(this);
        // 设置网页刷新监听
        mRefreshLayout.setOnRefreshListener(this);
    }

    @Override
    protected void initData() {
        showLoading();

        mBrowserView.setBrowserViewClient(new AppBrowserViewClient());
        mBrowserView.setBrowserChromeClient(new AppBrowserChromeClient(mBrowserView));
        boolean feedback = getBoolean(IntentKey.OTHER);
        if (feedback) {
            String openId = getString(IntentKey.ID);
            String nickName = getString(IntentKey.NAME);
            String avatar = getString(IntentKey.AVATAR_URL);
            Timber.d("initData：===> openId is " + openId + " nickName is " + nickName + " avatar is " + avatar);
            String postData = "nickname=" + nickName + "&avatar=" + avatar + "&openid=" + openId;
            mBrowserView.postUrl(getString(IntentKey.URL), postData.getBytes());
        } else {
            mBrowserView.loadUrl(getString(IntentKey.URL));
        }
    }

    @Override
    public StatusLayout getStatusLayout() {
        return mStatusLayout;
    }

    @Override
    public void onLeftClick(View view) {
        finish();
    }

    @Override
    public void onRightClick(View view) {
        UMWeb content = new UMWeb(mBrowserView.getUrl());
        content.setTitle(mBrowserView.getTitle());
        content.setThumb(new UMImage(this, R.mipmap.launcher_ic));
        content.setDescription(getString(R.string.app_name));
        // 分享
        new ShareDialog.Builder(this)
                .setShareLink(content)
                .setListener(new UmengShare.OnShareListener() {

                    @Override
                    public void onSucceed(Platform platform) {
                        toast("分享成功");
                    }

                    @Override
                    public void onError(Platform platform, Throwable t) {
                        toast(t.getMessage());
                    }

                    @Override
                    public void onCancel(Platform platform) {
                        toast("分享取消");
                    }
                })
                .show();
    }

    @Override
    public boolean onKeyDown(int keyCode, KeyEvent event) {
        if (keyCode == KeyEvent.KEYCODE_BACK && mBrowserView.canGoBack()) {
            // 后退网页并且拦截该事件
            mBrowserView.goBack();
            return true;
        }
        return super.onKeyDown(keyCode, event);
    }

    /**
     * 重新加载当前页
     */
    @CheckNet
    private void reload() {
        mBrowserView.reload();
    }

    /**
     * {@link OnRefreshListener}
     */

    @Override
    public void onRefresh(@NonNull RefreshLayout refreshLayout) {
        reload();
    }

    @Override
    protected boolean isStatusBarDarkFont() {
        return false;
    }

    private class AppBrowserViewClient extends BrowserView.BrowserViewClient {

        private final CookieRoomDatabase mDatabase = AppApplication.getDatabase();
        private final CookieDao mCookieDao = mDatabase.cookieDao();

        /**
         * 网页加载错误时回调，这个方法会在 onPageFinished 之前调用
         */
        @Override
        public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
            // 这里为什么要用延迟呢？因为加载出错之后会先调用 onReceivedError 再调用 onPageFinished
            post(() -> showError(v -> reload()));
        }

        /**
         * 开始加载网页
         */
        @Override
        public void onPageStarted(WebView view, String url, Bitmap favicon) {
            mProgressBar.setVisibility(View.VISIBLE);
            injectCookie(url);
        }

        private void injectCookie(String url) {
            String domain = StringUtil.getTopDomain(Constants.SUNNY_BEACH_API_BASE_URL);
            ThreadPoolManager manager = ThreadPoolManager.getInstance();
            manager.execute(() -> {
                Timber.d("hookUrlLoad：===> domain is %s", domain);
                CookieManager cookieManager = CookieManager.getInstance();
                CookieStore cookieStore = mCookieDao.getCookiesByDomain(domain);
                if (cookieStore != null) {
                    List<Cookie> cookieStoreList = cookieStore.getCookies();
                    for (Cookie cookie : cookieStoreList) {
                        String cookieName = cookie.name();
                        String cookieValue = cookie.value();
                        String cookieDomain = cookie.domain();
                        String cookieStr = new Cookie.Builder()
                                .name(cookieName)
                                .value(cookieValue)
                                .domain(cookieDomain)
                                .path("/")
                                .build()
                                .toString();
                        Timber.d("hookUrlLoad：===> Set-Cookie is %s", cookieStr);
                        cookieManager.setCookie(url, cookieStr);
                    }
                }
                String newCookie = cookieManager.getCookie(url);
                if (newCookie != null) {
                    Timber.d("hookUrlLoad：===> newCookie is %s", newCookie);
                }
                String currUrlTopDomain = StringUtil.getTopDomain(url);
                String apiTopDomain = StringUtil.getTopDomain(Constants.SUNNY_BEACH_API_BASE_URL);
                String siteTopDomain = StringUtil.getTopDomain(Constants.SUNNY_BEACH_SITE_BASE_URL);
                if (currUrlTopDomain.equals(apiTopDomain) || currUrlTopDomain.equals(siteTopDomain)) {
                    String cookieName = SobCacheManager.SOB_TOKEN_NAME;
                    String cookieValue = SobCacheManager.INSTANCE.getSobToken();
                    String apiCookie = new Cookie.Builder()
                            .name(cookieName)
                            .value(cookieValue)
                            .domain(apiTopDomain)
                            .path("/")
                            .build()
                            .toString();
                    String siteCookie = new Cookie.Builder()
                            .name(cookieName)
                            .value(cookieValue)
                            .domain(siteTopDomain)
                            .path("/")
                            .build()
                            .toString();
                    Timber.d("===> Set-Cookie：apiCookie is %s", apiCookie);
                    Timber.d("===> Set-Cookie：siteCookie is %s", siteCookie);
                    cookieManager.setCookie(url, apiCookie);
                    cookieManager.setCookie(url, siteCookie);
                }
                Timber.d("===> CookieManager is finish");
            });
        }

        /**
         * 完成加载网页
         */
        @Override
        public void onPageFinished(WebView view, String url) {
            mProgressBar.setVisibility(View.GONE);
            if (url.contains(Constants.SUNNY_BEACH_ARTICLE_URL_PRE)) {
                FitScreen fitScreen = new FitScreen(view);
                fitScreen.run();
            }
            mRefreshLayout.finishRefresh();
            showComplete();
        }
    }

    private class AppBrowserChromeClient extends BrowserView.BrowserChromeClient {

        private AppBrowserChromeClient(BrowserView view) {
            super(view);
        }

        /**
         * 收到网页标题
         */
        @Override
        public void onReceivedTitle(WebView view, String title) {
            if (title == null) {
                return;
            }
            setTitle(title);
        }

        @Override
        public void onReceivedIcon(WebView view, Bitmap icon) {
            if (icon == null) {
                return;
            }
            setRightIcon(new BitmapDrawable(getResources(), icon));
        }

        /**
         * 收到加载进度变化
         */
        @Override
        public void onProgressChanged(WebView view, int newProgress) {
            mProgressBar.setProgress(newProgress);
        }
    }
}

File: app/src/main/java/cn/cqautotest/sunnybeach/ui/activity/BrowserActivity.kt
package cn.cqautotest.sunnybeach.ui.activity

import android.annotation.SuppressLint
import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.BitmapDrawable
import android.net.Uri
import android.text.TextUtils
import android.view.MotionEvent
import android.view.View
import android.view.ViewConfiguration
import android.webkit.ConsoleMessage
import android.webkit.WebView
import android.widget.ProgressBar
import cn.cqautotest.sunnybeach.R
import cn.cqautotest.sunnybeach.action.StatusAction
import cn.cqautotest.sunnybeach.aop.Log
import cn.cqautotest.sunnybeach.app.AppActivity
import cn.cqautotest.sunnybeach.ktx.startActivity
import cn.cqautotest.sunnybeach.ui.dialog.ShareDialog
import cn.cqautotest.sunnybeach.util.SUNNY_BEACH_ARTICLE_URL_PRE
import cn.cqautotest.sunnybeach.util.SUNNY_BEACH_FISH_URL_PRE
import cn.cqautotest.sunnybeach.util.SUNNY_BEACH_QA_URL_PRE
import cn.cqautotest.sunnybeach.util.SUNNY_BEACH_SHARE_URL_PRE
import cn.cqautotest.sunnybeach.util.WebViewHookHelper
import cn.cqautotest.sunnybeach.widget.BrowserView
import cn.cqautotest.sunnybeach.widget.BrowserView.BrowserChromeClient
import cn.cqautotest.sunnybeach.widget.BrowserView.BrowserViewClient
import cn.cqautotest.sunnybeach.widget.StatusLayout
import com.flyjingfish.android_aop_core.annotations.CheckNetwork
import com.hjq.bar.TitleBar
import com.hjq.umeng.Platform
import com.hjq.umeng.UmengShare.OnShareListener
import com.scwang.smart.refresh.layout.SmartRefreshLayout
import com.scwang.smart.refresh.layout.api.RefreshLayout
import com.scwang.smart.refresh.layout.listener.OnRefreshListener
import com.umeng.socialize.media.UMImage
import com.umeng.socialize.media.UMWeb
import timber.log.Timber
import kotlin.math.abs

/**
 *    author : Android 轮子哥 & A Lonely Cat
 *    github : https://github.com/getActivity/AndroidProject-Kotlin
 *    time   : 2018/10/18
 *    desc   : 浏览器界面
 */
class BrowserActivity : AppActivity(), StatusAction, OnRefreshListener {

    companion object {

        private const val INTENT_KEY_IN_URL: String = "url"

        private const val OPEN_ID: String = "id"
        private const val NICK_NAME = "nickName"
        private const val AVATAR_URL = "avatarUrl"

        // 是否加载为反馈界面
        private const val IS_FEED_BACK = "isFeedback"

        @CheckNetwork(invokeListener = true)
        @Log
        fun start(
            context: Context,
            url: String,
            isFeedback: Boolean = false,
            openId: String = "",
            nickName: String = "",
            avatar: String = ""
        ) {
            if (TextUtils.isEmpty(url)) {
                return
            }
            context.startActivity<BrowserActivity> {
                putExtra(INTENT_KEY_IN_URL, url)
                putExtra(IS_FEED_BACK, isFeedback)
                putExtra(OPEN_ID, openId)
                putExtra(NICK_NAME, nickName)
                putExtra(AVATAR_URL, avatar)
            }
        }
    }

    private val hintLayout: StatusLayout? by lazy { findViewById(R.id.hl_browser_hint) }
    private val progressBar: ProgressBar? by lazy { findViewById(R.id.pb_browser_progress) }
    private val refreshLayout: SmartRefreshLayout? by lazy { findViewById(R.id.sl_browser_refresh) }
    private val browserView: BrowserView? by lazy { findViewById(R.id.wv_browser_view) }

    override fun getLayoutId(): Int = R.layout.browser_activity

    override fun initView() {
        // 设置 WebView 生命管控
        browserView?.setLifecycleOwner(this)
        // 设置网页刷新监听
        refreshLayout?.setOnRefreshListener(this)
    }

    override fun initData() {
        showLoading()
        browserView?.apply {
            setBrowserViewClient(AppBrowserViewClient())
            setBrowserChromeClient(AppBrowserChromeClient(this))
            val url = getString(INTENT_KEY_IN_URL)!!
            val isFeedback = getBoolean(IS_FEED_BACK)
            val newUrl = if (isFeedback) {
                val openId = getString(OPEN_ID)
                val nickName = getString(NICK_NAME)
                val avatar = getString(AVATAR_URL)
                Timber.d("initData：===> openId is $openId nickName is $nickName avatar is $avatar")
                Uri.parse(url)
                    .buildUpon()
                    .appendQueryParameter("nickname", nickName ?: "游客")
                    .appendQueryParameter("avatar", avatar.orEmpty())
                    .appendQueryParameter("openid", openId.orEmpty())
                    .toString()
            } else {
                url
            }
            Timber.d("initData：===> newUrl is $newUrl")
            loadUrl(newUrl)
        }
    }

    @SuppressLint("ClickableViewAccessibility")
    override fun initEvent() {
        val clickTap = ViewConfiguration.getDoubleTapTimeout()
        // 最低滑动距离
        val minDist = ViewConfiguration.get(this).scaledTouchSlop
        var lastClickTime = 0L
        var x = 0f
        var y = 0f
        browserView?.setOnTouchListener { _, event ->
            when (event.action and MotionEvent.ACTION_MASK) {
                MotionEvent.ACTION_DOWN -> {
                    lastClickTime = System.currentTimeMillis()
                    x = event.x
                    y = event.y
                    false
                }

                MotionEvent.ACTION_UP -> {
                    val singClick = System.currentTimeMillis() - lastClickTime < clickTap
                    val xDist = abs(event.x - x)
                    val yDist = abs(event.y - y)
                    takeIf { interceptResClick() && singClick && xDist < minDist && yDist < minDist }?.let { handleResClick() } ?: false
                }

                else -> false
            }
        }
    }

    /**
     * 是否拦截处理
     */
    private fun interceptResClick(): Boolean {
        val curUrl = browserView?.url ?: return false
        return when {
            curUrl.startsWith(SUNNY_BEACH_ARTICLE_URL_PRE) -> true
            curUrl.startsWith(SUNNY_BEACH_FISH_URL_PRE) -> true
            curUrl.startsWith(SUNNY_BEACH_QA_URL_PRE) -> true
            curUrl.startsWith(SUNNY_BEACH_SHARE_URL_PRE) -> true
            else -> false
        }
    }

    /**
     * 处理资源被点击的情况
     */
    private fun handleResClick(): Boolean {
        val hitTestResult = browserView?.hitTestResult ?: return false
        return when (hitTestResult.type) {
            WebView.HitTestResult.IMAGE_TYPE, WebView.HitTestResult.IMAGE_ANCHOR_TYPE -> {
                try {
                    handleImage(hitTestResult.extra)
                } catch (t: Throwable) {
                    t.printStackTrace()
                    // 处理图片错误，不支持查看的图片类型（可能是 base64 的图片，图片过大）
                    return false
                }
                true
            }

            else -> false
        }
    }

    /**
     * 处理图片被单击的情况
     */
    private fun handleImage(imageLink: String?) {
        imageLink?.let { ImagePreviewActivity.start(this, it) }
    }

    override fun getStatusLayout(): StatusLayout? = hintLayout

    override fun onLeftClick(titleBar: TitleBar) {
        finish()
    }

    override fun onRightClick(titleBar: TitleBar) {
        val content = UMWeb(browserView?.url)
        content.title = browserView?.title
        content.setThumb(UMImage(this, R.mipmap.launcher_ic))
        content.description = getString(R.string.app_name)
        // 分享
        ShareDialog.Builder(this)
            .setShareLink(content)
            .setListener(object : OnShareListener {
                override fun onSucceed(platform: Platform?) {
                    toast("分享成功")
                }

                override fun onError(platform: Platform?, t: Throwable) {
                    toast(t.message)
                }

                override fun onCancel(platform: Platform?) {
                    toast("分享取消")
                }
            })
            .show()
    }

    override fun onBackPressed() {
        browserView.takeIf { it != null && it.canGoBack() }?.goBack() ?: super.onBackPressed()
    }

    /**
     * 重新加载当前页
     */
    @CheckNetwork(invokeListener = true)
    private fun reload() {
        browserView?.reload()
    }

    /**
     * [OnRefreshListener]
     */
    override fun onRefresh(refreshLayout: RefreshLayout) {
        reload()
    }

    override fun isStatusBarDarkFont() = false

    private inner class AppBrowserViewClient : BrowserViewClient() {

        /**
         * 网页加载错误时回调，这个方法会在 onPageFinished 之前调用
         */
        override fun onReceivedError(view: WebView, errorCode: Int, description: String, failingUrl: String) {
            Timber.d("onReceivedError：===> errorCode is $errorCode description is $description failingUrl is $failingUrl")
            // 这里为什么要用延迟呢？因为加载出错之后会先调用 onReceivedError 再调用 onPageFinished
            post {
                showError { reload() }
            }
        }

        /**
         * 开始加载网页
         */
        override fun onPageStarted(view: WebView, url: String, favicon: Bitmap?) {
            progressBar?.visibility = View.VISIBLE
            // 给指定的 url 注入 cookie
            WebViewHookHelper.injectCookie(url)
        }

        /**
         * 完成加载网页
         */
        override fun onPageFinished(view: WebView, url: String) {
            progressBar?.visibility = View.GONE
            WebViewHookHelper.fitScreen(view, url)
            refreshLayout?.finishRefresh()
            showComplete()
        }
    }

    private inner class AppBrowserChromeClient constructor(view: BrowserView) : BrowserChromeClient(view) {

        override fun onConsoleMessage(consoleMessage: ConsoleMessage?): Boolean {
            Timber.d("onConsoleMessage：===> " + consoleMessage?.message())
            return true
        }

        /**
         * 收到网页标题
         */
        override fun onReceivedTitle(view: WebView, title: String?) {
            title?.let { setTitle(title) }
        }

        override fun onReceivedIcon(view: WebView, icon: Bitmap?) {
            icon?.let { setRightIcon(BitmapDrawable(resources, icon)) }
        }

        /**
         * 收到加载进度变化
         */
        override fun onProgressChanged(view: WebView, newProgress: Int) {
            progressBar?.progress = newProgress
        }
    }
}